src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/course/domain/CourseService.kt

```diff
@@ -75,28 +75,51 @@ class CourseService(
     courses.forEach { it.prerequisites.clear() }
   }
 
-  fun replaceAllOfferings(replacements: List<NewOffering>): List<LineMessage> {
+  fun updateOfferings(updates: List<OfferingUpdate>): List<LineMessage> {
     val allCourses = courseRepository.allCourses()
-    clearOfferings(allCourses)
     val coursesByIdentifier = allCourses.associateBy(CourseEntity::identifier)
-    replacements.forEach { record ->
-      coursesByIdentifier[record.identifier]?.run {
-        addOffering(
-          Offering(
-            organisationId = record.prisonId,
-            contactEmail = record.contactEmail
-              ?: "",
-            secondaryContactEmail = record.secondaryContactEmail,
-          ),
-        )
+    val desiredOfferingsByCourseIdentifier = updates.groupBy(OfferingUpdate::identifier)
+    coursesByIdentifier.forEach { (courseIdentifier, course) ->
+      updateOfferingsForCourse(course, desiredOfferingsByCourseIdentifier[courseIdentifier] ?: emptyList())
+    }
+
+    return contactEmailWarnings(updates) + unmatchedCourseErrors(updates, coursesByIdentifier)
+  }
+
+  private fun updateOfferingsForCourse(course: CourseEntity, desiredOfferings: List<OfferingUpdate>) {
```

Comment:
This method is doing a few things, I wonder if it would be more easily scannable for future readers if we split it into three smaller private methods:
`addNewOfferingsToCourse`
`updateCourseOfferings`
`withdrawOfferingsFromCourse`

What do you think?